lib/diaspora/federation/receive.rb

```diff
@@ -0,0 +1,322 @@
+      end
+
+      def self.participation(entity)
+        author = author_of(entity)
+        ignore_existing_guid(Participation, entity.guid, author) do
+          parent = entity.parent_type.constantize.find_by(guid: entity.parent_guid)
+
+          Participation.create!(author: author, guid: entity.guid, target: parent) if parent.author.local?
+        end
+      end
+
+      def self.photo(entity)
+        author = author_of(entity)
+        persisted_photo = load_from_db(Photo, entity.guid, author)
+
+        if persisted_photo
+          persisted_photo.tap do |photo|
+            photo.update_attributes(
+              text:                entity.text,
+              public:              entity.public,
+              created_at:          entity.created_at,
+              remote_photo_path:   entity.remote_photo_path,
+              remote_photo_name:   entity.remote_photo_name,
+              status_message_guid: entity.status_message_guid,
+              height:              entity.height,
+              width:               entity.width
+            )
+          end
+        else
+          save_photo(entity)
+        end
+      end
+
+      def self.poll_participation(entity)
+        receive_relayable(PollParticipation, entity) do
+          PollParticipation.new(
+            author:           author_of(entity),
+            guid:             entity.guid,
+            poll:             Poll.find_by(guid: entity.parent_guid),
+            poll_answer_guid: entity.poll_answer_guid
+          )
+        end
+      end
+
+      def self.profile(entity)
+        author_of(entity).profile.tap do |profile|
+          profile.update_attributes(
+            first_name:       entity.first_name,
+            last_name:        entity.last_name,
+            image_url:        entity.image_url,
+            image_url_medium: entity.image_url_medium,
+            image_url_small:  entity.image_url_small,
+            birthday:         entity.birthday,
+            gender:           entity.gender,
+            bio:              entity.bio,
+            location:         entity.location,
+            searchable:       entity.searchable,
+            nsfw:             entity.nsfw,
+            tag_string:       entity.tag_string
+          )
+        end
+      end
+
+      def self.reshare(entity)
+        author = author_of(entity)
+        ignore_existing_guid(Reshare, entity.guid, author) do
+          Reshare.create!(
+            author:                author,
+            guid:                  entity.guid,
+            created_at:            entity.created_at,
+            provider_display_name: entity.provider_display_name,
+            public:                entity.public,
+            root_guid:             entity.root_guid
+          )
+        end
+      end
+
+      def self.retraction(entity, recipient_id)
+        object = entity.target_type.constantize.where(guid: entity.target_guid).take!
```

Comment:
Yes, the federation-gem already checks the target on receive, and checks if the sender of the retraction is allowed to delete this entity, otherwise it would raise a `DiasporaFederation::Entities::Retraction::TargetNotFound` or `DiasporaFederation::Federation::Receiver::InvalidSender`